### PR TITLE
add configuration to auto copy file remotely

### PR DESCRIPTION
**Pull request by yungezz**
## Summary

add new feature:
copy local saved files to remote ssh host.

## Issue Type

- New Feature

### CODE REVIEW

File: src/fileSyncer.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    public updateConfiguration(config: any): void {
+        // check if changed
+        let updatedConfig = this.hasConfigurationChanged(this._configuration, config);
+
+        this.copyFiles(updatedConfig);
+
+        this._configuration = config;
+    }
+
+    protected hasConfigurationChanged(oldConfig: FileCopyConfigs, newConfig: FileCopyConfigs) {
```
Comment (yuwzho) [inline src/fileSyncer.ts]: This function name sounds like a function to detect whether there is change happens and the return type should be bool.

File: src/fileSyncer.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        let result = [];
+
+        if (!oldConfig || oldConfig.length === 0) {
```
Comment (yuwzho) [inline src/fileSyncer.ts]: this conditional check can be written as `oldConfig = oldConfig || []`. This empty check should also be applied to newConfig. And then go to the original logic to merge two array

File: src/fileSyncer.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            }
+        }
+
+        return result;
+    }
+
+    public async copyFiles(configuration: FileCopyConfigs, fileName: string = null) {
+        let servers = utilities.getSSHConfig();
+
+        if (!configuration) {
```
Comment (yuwzho) [inline src/fileSyncer.ts]: `configuration = configuration || this._configuration || []`

File: src/fileSyncer.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+            if (!exists) {
+                result.push(newc);
+            }
+        }
+
+        return result;
+    }
+
+    public async copyFiles(configuration: FileCopyConfigs, fileName: string = null) {
```
Comment (yuwzho) [inline src/fileSyncer.ts]: where is the await?

File: src/sshRunner.ts [hunk 5]
```diff
@@ -70,51 +71,80 @@ export class SSHRunner extends TerminalBaseRunner {
...
         let target = path.join('\./', path.basename(playbook));
         let targetPlaybook = target;
 
+        // check configuration
+        let fileSyncConfig = utilities.getCodeConfiguration<FileCopyConfigs>('ansible', 'fileCopyConfig');
 
-        // check configuration weather to sync workspace
-        let copyWorkspace = utilities.getCodeConfiguration<boolean>('ansible', 'copyWorkspace');
+        if (fileSyncConfig) {
+            for (let config of fileSyncConfig) {
```
Comment (yuwzho) [inline src/sshRunner.ts]: `for (let config of (fileSyncConfig || [])) {`

File: src/sshRunner.ts [hunk 6]
```diff
@@ -70,51 +71,80 @@ export class SSHRunner extends TerminalBaseRunner {
...
-            src = utilities.getWorkspaceRoot(playbook) + '/';
-            target = path.join('\./', path.basename(src)) + '/';
-            targetPlaybook = ['\./' + path.basename(src), path.relative(src, playbook)].join(path.posix.sep).replace(/\\/g, '/');
-
-            try {
-                await this.folderSyncer.syncFolder(src, target, targetServer, false);
-            } catch (err) {
-                return;
+                    // if not saved on copy, copy playbook to remote
+                    if (!config.copyOnSave) {
```
Comment (yuwzho) [inline src/sshRunner.ts]: what if 
1. copyOnSave = false
2. modified some file
3. copyOnSave = true
4. run

File: src/utilities.ts [hunk 7]
```diff
@@ -333,5 +343,22 @@ export function delayedInterval(func: () => void, interval: number) {
...
+}
+
+export function isSubPath(child: string, parent: string): boolean {
+    if (child === parent) return false;
+    let parentTokens = [];
+    parent.split(/\\/).forEach((s) => parentTokens = parentTokens.concat(s.split(/\//).filter(i => i.length)));
+
+    let childTokens = [];
+    child.split(/\\/).forEach((s) => childTokens = childTokens.concat(s.split(/\//).filter(i => i.length)));
+    return parentTokens.every((t, i) => childTokens[i] === t);
```
Comment (yuwzho) [inline src/utilities.ts]: Need wait the forEach terminate?

File: src/fileSyncer.ts (hunk 1 above)
Comment (yungezz) [reply src/fileSyncer.ts]: fixed

Comment (yungezz) [reply src/fileSyncer.ts]: fixed

File: src/fileSyncer.ts (hunk 2 above)
Comment (yungezz) [reply src/fileSyncer.ts]: fixed

File: src/fileSyncer.ts (hunk 4 above)
Comment (yungezz) [reply src/fileSyncer.ts]: removed async

File: src/sshRunner.ts (hunk 6 above)
Comment (yungezz) [reply src/sshRunner.ts]: run against old version.

**Pull request merged**